src/token/token.service.ts

```diff
@@ -55,11 +118,208 @@ export class TokenService {
       const tokenResults = await axios.get(priceServiceUrl);
       return tokenResults.data;
     } catch (e) {
-      this.logger.warn(`Data request error: ${e}`);
+      this.logger.warn(`Data request error from price service: ${e}`);
     }
   }
 
+  async requestNearPrice() {
+    try {
+      const nearHelperUrl = configService.getHelperUrl();
+      const nearData = await axios.get(`${nearHelperUrl}/fiat`);
+      return nearData.data.near.usd;
+    } catch (e) {
+      this.logger.warn(`Data request error from helper: ${e}`);
+    }
+  }
+
+  private async getPoolsPage(
+    service,
+    from: number,
+    limit: number = DEFAULT_PAGE_LIMIT,
+  ): Promise<Pool[]> {
+    try {
+      const poolPage = await service.viewFunction('get_pools', {
+        from_index: from,
+        limit: limit,
+      });
+      return poolPage.map((pool) => {
+        pool.supplies = pool.amounts.reduce(
+          (acc: { [tokenId: string]: string }, amount: string, i: number) => {
+            acc[pool.token_account_ids[i]] = amount;
+            return acc;
+          },
+          {},
+        );
+        return pool;
+      });
+    } catch (e) {
+      this.logger.warn(`Data request error from getting pool page: ${e}`);
+      return [];
+    }
+  }
+
+  async getDataFromPools() {
+    try {
+      const service = await initializeNearService();
```

Comment:
Initialization of Near instance should be done in the service constructor to avoid excessive memory allocations.

Reply:
How we can create initializing like that for async services?

Reply:
```typescript
Injectable()
export class TokenService {
  private readonly logger = new Logger(TokenService.name);
  private readonly near: ReturnType<typeof initializeNearService>,
  constructor(
    @InjectRepository(Token)
    private readonly tokenRepo: Repository<Token>,
  ) {
    this.near = await initializeNearService();
  }
...
```

Reply:
'await' expressions are only allowed within async functions and at the top levels of modules and also we can't do async constructor
what about next? 
```
private connection = null;

  constructor(
    @InjectRepository(Token)
    private readonly tokenRepo: Repository<Token>,
  ) {}

  private async getConnection() {
    if (!this.connection) {
      const connection = await initializeNearConnection();
      this.connection = connection;
      return connection;
    }
    return this.connection;
  }
```

Reply:
My apologies, I've missed the async part of the problem. Here's a canonical solution:

```typescript
@Injectable()
export class TokenService implements OnModuleInit {
  private readonly logger = new Logger(TokenService.name);
  private near: Awaited<ReturnType<typeof initializeNearService>>;

  constructor(
    @InjectRepository(Token)
    private readonly tokenRepo: Repository<Token>,
  ) {}

  async onModuleInit() {
    this.near = await initializeNearService();
  }
```